src/devices/machine/cs8900a.cpp

```diff
@@ -0,0 +1,1522 @@
+
+   REMARK: The code operatoes the cs8900a in IO space configuration, as
+   it generates I/OW and I/OR signals.
+*/
+
+/*
+    RW: RXTXDATA   = DE00/DE01
+    RW: RXTXDATA2  = DE02/DE03 (for 32-bit-operation)
+    -W: TXCMD      = DE04/DE05 (TxCMD, Transmit Command)   mapped to PP + 0144 (Reg. 9, Sec. 4.4, page 46)
+    -W: TXLENGTH   = DE06/DE07 (TxLenght, Transmit Length) mapped to PP + 0146 
+    R-: INTSTQUEUE = DE08/DE09 (Interrupt Status Queue)    mapped to PP + 0120 (ISQ, Sec. 5.1, page 78)
+    RW: PP_PTR     = DE0A/DE0B (PacketPage Pointer)        (see. page 75p: Read -011.---- ----.----)
+    RW: PP_DATA0   = DE0C/DE0D (PacketPage Data (Port 0))  
+    RW: PP_DATA1   = DE0E/DE0F (PacketPage Data (Port 1))  (for 32 bit only)
+*/
+
+#define CS8900_ADDR_RXTXDATA   0x00 /* RW */
+#define CS8900_ADDR_RXTXDATA2  0x02 /* RW 32 bit only! */
+#define CS8900_ADDR_TXCMD      0x04 /* -W Maps to PP+0144 */
+#define CS8900_ADDR_TXLENGTH   0x06 /* -W Maps to PP+0146 */
+#define CS8900_ADDR_INTSTQUEUE 0x08 /* R- Interrupt status queue, maps to PP + 0120 */
+#define CS8900_ADDR_PP_PTR     0x0a /* RW PacketPage Pointer */
+#define CS8900_ADDR_PP_DATA    0x0c /* RW PacketPage Data, Port 0 */
+#define CS8900_ADDR_PP_DATA2   0x0e /* RW PacketPage Data, Port 1 - 32 bit only */    
+
+/* Makros for reading and writing the visible register: */
+#define GET_CS8900_8(  _xxx_ ) \
+    ( assert(_xxx_<CS8900_COUNT_IO_REGISTER), \
+      cs8900_ioregs[_xxx_]                           \
+    )
+
+#define SET_CS8900_8( _xxx_, _val_ ) \
+    do { \
+        assert(_xxx_<CS8900_COUNT_IO_REGISTER); \
+        cs8900_ioregs[_xxx_  ] = (_val_     ) & 0xff; \
+    } while (0) 
+
+#define GET_CS8900_16(  _xxx_ ) \
+    ( assert(_xxx_<CS8900_COUNT_IO_REGISTER), \
+      cs8900_ioregs[_xxx_] | (cs8900_ioregs[_xxx_+1] << 8)     \
+    )
+
+#define SET_CS8900_16( _xxx_, _val_ ) \
+    do { \
+        assert(_xxx_<CS8900_COUNT_IO_REGISTER); \
+        cs8900_ioregs[_xxx_  ] = (_val_     ) & 0xff; \
+        cs8900_ioregs[_xxx_+1] = (_val_ >> 8) & 0xff; \
+    } while (0) 
+
+/* The PacketPage register */
+/* note: The locations 0 to MAX_PACKETPAGE_ARRAY-1 are handled in this array. */
+
+/* Makros for reading and writing the PacketPage register: */
+
+#define GET_PP_8(  _xxx_ ) \
+    (assert(_xxx_<MAX_PACKETPAGE_ARRAY), \
+     cs8900_packetpage[_xxx_] \
+    )
+
+#define GET_PP_16(  _xxx_ ) \
+    ( assert(_xxx_<MAX_PACKETPAGE_ARRAY),   \
+      assert((_xxx_ & 1) == 0 ),            \
+      ((u16)cs8900_packetpage[_xxx_]        ) \
+    | ((u16)cs8900_packetpage[_xxx_+1] <<  8) \
+    )
+
+#define GET_PP_32(  _xxx_ ) \
+    ( assert(_xxx_<MAX_PACKETPAGE_ARRAY),     \
+      assert((_xxx_ & 3) == 0 ),              \
+      (((long)cs8900_packetpage[_xxx_  ])      ) \
+    | (((long)cs8900_packetpage[_xxx_+1]) <<  8) \
+    | (((long)cs8900_packetpage[_xxx_+2]) << 16) \
+    | (((long)cs8900_packetpage[_xxx_+3]) << 24) \
```

Comment:
Don’t use `long` unless you have a very good reason to – I think you want to use `u32` here.